File: Numeros Pseudo-Aleatorios/Mersenne Twister/MersenneTwister.scala
/**
 * PRNG .......: Mersenne Twister (MT)
 * Ano ........: 1998
 * Autor ......: Makoto Matsumoto & Takuji Nishimura
 */

object Main {
  def main(args: Array[String]): Unit = {

    class MT19937 {
      private val w: Int = 32
      private val n: Int = 624
      private val f: Int = 1812433253
      private val m: Int = 397
      private val r: Int = 31
      private val a: Int = 0x9908B0DF
      private val d: Int = 0xFFFFFFFF.toInt
      private val b: Int = 0x9D2C5680
      private val c: Int = 0xEFC60000
      private val u: Int = 11
      private val s: Int = 7
      private val t: Int = 15
      private val l: Int = 18

      private var X: Array[Int] = new Array[Int](n)
      private var cnt: Int = 0

      private def twist(): Unit = {
        for (i <- 0 until n) {
          val lower_mask: Int = (1 << r) - 1
          val upper_mask: Int = ~lower_mask
          val tmp: Int = (X(i) & upper_mask) + (X((i + 1) % n) & lower_mask)
          val tmpA: Int = tmp >> 1
          if ((tmp & 1) != 0) {
            tmpA ^ a
          }
          X(i) = X((i + m) % n) ^ tmpA
        }
        cnt = 0
      }

      def initialize(seed: Int): Unit = {
        X(0) = seed
        for (i <- 1 until n) {
          X(i) = (f * (X(i - 1) ^ (X(i - 1) >> (w - 2))) + i) & 0xFFFFFFFF
        }
        twist()
      }

      def generateRandomNumber(): Int = {
        if (cnt == n) {
          twist()
        }
        var y: Int = X(cnt)
        y ^= (y >> u) & d
        y ^= (y << s) & b
        y ^= (y << t) & c
        y ^= y >> l
        cnt += 1
        y
      }
    }

    val Xn1: MT19937 = new MT19937
    val Xn2: MT19937 = new MT19937

    Xn1.initialize(1234)
    Xn2.initialize(System.currentTimeMillis().toInt)

    println("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)")
    println("PSEUDORANDOM NUMBER GENERATOR (PRNG)")
    println("PADRAO: Mersenne Twister (MT)")
    println("==========================================\n")
    println("    Semente: 1234        Semente: TIME")
    println("    -------------        -------------\n")

    for (_ <- 0 to 9) {
      print(f"${Xn1.generateRandomNumber()}%17s")
      print(f"${Xn2.generateRandomNumber()}%21s")
      println()
    }
  }
}

File: Numeros Pseudo-Aleatorios/Mersenne Twister/MersenneTwister.java
/**
 * PRNG .......: Mersenne Twister (MT)
 * Ano ........: 1998
 * Autor ......: Makoto Matsumoto & Takuji Nishimura
 */

public class Main {
	
    public static class MT19937 {
        private int w = 32;
        private int n = 624;
        private int f = 1812433253;
        private int m = 397;
        private int r = 31;
        private int a = 0x9908B0DF;
        private int d = 0xFFFFFFFF;
        private int b = 0x9D2C5680;
        private int c = 0xEFC60000;
        private int u = 11;
        private int s = 7;
        private int t = 15;
        private int l = 18;

        private int[] X = new int[n];
        private int cnt = 0;

        private void twist() {
            for (int i = 0; i < n; i++) {
                int lower_mask = (1 << r) - 1;
                int upper_mask = ~lower_mask;
                int tmp = (X[i] & upper_mask) + (X[(i + 1) % n] & lower_mask);
                int tmpA = tmp >> 1;
                if ((tmp & 1) != 0) {
                    tmpA ^= a;
                }
                X[i] = X[(i + m) % n] ^ tmpA;
            }
            cnt = 0;
        }

        public void initialize(int seed) {
            X[0] = seed;
            for (int i = 1; i < n; i++) {
                X[i] = (f * (X[i - 1] ^ (X[i - 1] >> (w - 2))) + i) & 0xFFFFFFFF;
            }
            twist();
        }

        public int generateRandomNumber() {
            if (cnt == n) {
                twist();
            }
            int y = X[cnt];
            y ^= (y >> u) & d;
            y ^= (y << s) & b;
            y ^= (y << t) & c;
            y ^= y >> l;
            cnt++;
            return y;
        }
    }

    public static void main(String[] args) {
		
        MT19937 Xn1 = new MT19937();
        MT19937 Xn2 = new MT19937();

        Xn1.initialize(1234);
        Xn2.initialize((int) System.currentTimeMillis());

        System.out.println("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)");
        System.out.println("PSEUDORANDOM NUMBER GENERATOR (PRNG)");
        System.out.println("PADRAO: Mersenne Twister (MT)");
        System.out.println("==========================================\n");
        System.out.println("    Semente: 1234        Semente: TIME");
        System.out.println("    -------------        -------------\n");

        for (int i = 0; i <= 9; i++) {
            System.out.printf("%17s", Xn1.generateRandomNumber());
            System.out.printf("%21s\n", Xn2.generateRandomNumber());
        }
        
    }
    
}
